calico/felix/fetcd.py

```diff
@@ -115,13 +118,25 @@ def __init__(self, config, hosts_ipset):
         self._resync_greenlet = gevent.spawn(self._periodically_resync)
         self._resync_greenlet.link_exception(self._on_worker_died)
 
-        # Start up a reporting greenlet.
+        # Start up greenlet to report felix's liveness into etcd.
         self.done_first_status_report = False
         self._status_reporting_greenlet = gevent.spawn(
             self._periodically_report_status
         )
         self._status_reporting_greenlet.link_exception(self._on_worker_died)
 
+        # Start up a greenlet to report per-endpoint status into etcd.
+        self._endpoint_status = {}
+        self._dirty_endpoints = set()
+        self._endpoint_status_lock = Semaphore()
+        self._endpoint_status_work_to_do = Event()
+        self._per_endpoint_reporting_greenlet = gevent.spawn(
```

Comment:
Do Semaphores and events play nicely with gevent? I assume they do, but it felt like we should have stuck to the actor model here, since the greenlet that handles these events could become an actor, and then wouldn't things work just the same?